H2DB file corruption

Tested in Kura 3.3.0-SNAPSHOT but valid for all the versions since the introduction of the bundle.
When the database is in most persistent setup and a power loss happens, it could be that at the next boot the database file is corrupted.

In this case, an exception like this is thrown:
```
2018-07-25 17:04:50,079 [Start Level: Equinox Container: e3729c13-30bd-4b97-9aa6-783d41607f14] INFO  o.e.k.c.d.H2DbServiceImpl - Opening database with url: jdbc:h2:file:/home/root/db
2018-07-25 17:04:51,053 [Start Level: Equinox Container: e3729c13-30bd-4b97-9aa6-783d41607f14] ERROR o.e.k.c.d.H2DbServiceImpl - Error getting connection
org.h2.jdbc.JdbcSQLException: File corrupted while reading record: null. Possible solution: use the recovery tool [90030-192]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:168)
	at org.h2.mvstore.db.MVTableEngine$Store.convertIllegalStateException(MVTableEngine.java:195)
	at org.h2.mvstore.db.MVTableEngine$Store.open(MVTableEngine.java:167)
	at org.h2.mvstore.db.MVTableEngine.init(MVTableEngine.java:99)
	at org.h2.engine.Database.getPageStore(Database.java:2463)
	at org.h2.engine.Database.open(Database.java:692)
	at org.h2.engine.Database.openDatabase(Database.java:270)
	at org.h2.engine.Database.<init>(Database.java:264)
	at org.h2.engine.Engine.openSession(Engine.java:64)
	at org.h2.engine.Engine.openSession(Engine.java:176)
	at org.h2.engine.Engine.createSessionAndValidate(Engine.java:154)
	at org.h2.engine.Engine.createSession(Engine.java:137)
	at org.h2.engine.Engine.createSession(Engine.java:27)
	at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:349)
	at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:118)
	at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:102)
	at org.h2.Driver.connect(Driver.java:72)
	at org.h2.jdbcx.JdbcDataSource.getJdbcConnection(JdbcDataSource.java:190)
	at org.h2.jdbcx.JdbcDataSource.getXAConnection(JdbcDataSource.java:353)
	at org.h2.jdbcx.JdbcDataSource.getPooledConnection(JdbcDataSource.java:385)
	at org.h2.jdbcx.JdbcConnectionPool.getConnectionNow(JdbcConnectionPool.java:227)
	at org.h2.jdbcx.JdbcConnectionPool.getConnection(JdbcConnectionPool.java:199)
	at org.eclipse.kura.core.db.H2DbServiceImpl.getConnection(H2DbServiceImpl.java:177)
	at org.eclipse.kura.core.db.H2DbServiceImpl.openDatabase(H2DbServiceImpl.java:321)
	at org.eclipse.kura.core.db.H2DbServiceImpl.openConnectionPool(H2DbServiceImpl.java:315)
	at org.eclipse.kura.core.db.H2DbServiceImpl.updated(H2DbServiceImpl.java:126)
	at org.eclipse.kura.core.db.H2DbServiceImpl.activate(H2DbServiceImpl.java:89)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.apache.felix.scr.impl.inject.BaseMethod.invokeMethod(BaseMethod.java:229)
	at org.apache.felix.scr.impl.inject.BaseMethod.access$500(BaseMethod.java:39)
	at org.apache.felix.scr.impl.inject.BaseMethod$Resolved.invoke(BaseMethod.java:650)
	at org.apache.felix.scr.impl.inject.BaseMethod.invoke(BaseMethod.java:506)
	at org.apache.felix.scr.impl.inject.ActivateMethod.invoke(ActivateMethod.java:307)
	at org.apache.felix.scr.impl.inject.ActivateMethod.invoke(ActivateMethod.java:299)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:298)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:109)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:906)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:879)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:823)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse$1.run(ServiceFactoryUse.java:212)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.factoryGetService(ServiceFactoryUse.java:210)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.getService(ServiceFactoryUse.java:111)
	at org.eclipse.osgi.internal.serviceregistry.ServiceConsumer$2.getService(ServiceConsumer.java:45)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.getService(ServiceRegistrationImpl.java:508)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.getService(ServiceRegistry.java:461)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.getService(BundleContextImpl.java:624)
	at com.eclipsesource.jaxrs.publisher.internal.ResourceTracker.addingService(ResourceTracker.java:39)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:941)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:1)
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
	at org.osgi.util.tracker.AbstractTracked.trackInitial(AbstractTracked.java:183)
	at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:318)
	at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:261)
	at com.eclipsesource.jaxrs.publisher.internal.Activator.openAllServiceTracker(Activator.java:91)
	at com.eclipsesource.jaxrs.publisher.internal.Activator.start(Activator.java:55)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:779)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:1)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:772)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:729)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:933)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:309)
	at org.eclipse.osgi.container.Module.doStart(Module.java:581)
	at org.eclipse.osgi.container.Module.start(Module.java:449)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1634)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1614)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1585)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1528)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340)
Caused by: java.lang.IllegalStateException: File corrupted in chunk 15087, expected page length 4..768, got 200902975 [1.4.192/6]
	at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:773)
	at org.h2.mvstore.Page.read(Page.java:649)
	at org.h2.mvstore.Page.read(Page.java:195)
	at org.h2.mvstore.MVStore.readPage(MVStore.java:1944)
	at org.h2.mvstore.MVMap.readPage(MVMap.java:736)
	at org.h2.mvstore.Page.getChildPage(Page.java:217)
	at org.h2.mvstore.MVMap.binarySearch(MVMap.java:468)
	at org.h2.mvstore.MVMap.binarySearch(MVMap.java:469)
	at org.h2.mvstore.MVMap.get(MVMap.java:450)
	at org.h2.mvstore.MVStore.getChunkIfFound(MVStore.java:956)
	at org.h2.mvstore.MVStore.getChunk(MVStore.java:933)
	at org.h2.mvstore.MVStore.readPage(MVStore.java:1935)
	at org.h2.mvstore.MVMap.readPage(MVMap.java:736)
	at org.h2.mvstore.Page.getChildPage(Page.java:217)
	at org.h2.mvstore.MVMap.binarySearch(MVMap.java:468)
	at org.h2.mvstore.MVMap.binarySearch(MVMap.java:469)
	at org.h2.mvstore.MVMap.get(MVMap.java:450)
	at org.h2.mvstore.MVStore.getChunkIfFound(MVStore.java:956)
	at org.h2.mvstore.MVStore.getChunk(MVStore.java:933)
	at org.h2.mvstore.MVStore.readPage(MVStore.java:1935)
	at org.h2.mvstore.MVMap.readPage(MVMap.java:736)
	at org.h2.mvstore.Page.getChildPage(Page.java:217)
	at org.h2.mvstore.Cursor.min(Cursor.java:129)
	at org.h2.mvstore.Cursor.hasNext(Cursor.java:36)
	at org.h2.mvstore.MVStore.loadChunkMeta(MVStore.java:691)
	at org.h2.mvstore.MVStore.readStoreHeader(MVStore.java:672)
	at org.h2.mvstore.MVStore.<init>(MVStore.java:355)
	at org.h2.mvstore.MVStore$Builder.open(MVStore.java:2893)
	at org.h2.mvstore.db.MVTableEngine$Store.open(MVTableEngine.java:154)
	... 72 more
2018-07-25 17:04:51,066 [Start Level: Equinox Container: e3729c13-30bd-4b97-9aa6-783d41607f14] ERROR o.e.k.c.d.H2DbServiceImpl - Failed to open database
org.h2.jdbc.JdbcSQLException: File corrupted while reading record: null. Possible solution: use the recovery tool [90030-192]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:168)
	at org.h2.mvstore.db.MVTableEngine$Store.convertIllegalStateException(MVTableEngine.java:195)
	at org.h2.mvstore.db.MVTableEngine$Store.open(MVTableEngine.java:167)
	at org.h2.mvstore.db.MVTableEngine.init(MVTableEngine.java:99)
	at org.h2.engine.Database.getPageStore(Database.java:2463)
	at org.h2.engine.Database.open(Database.java:692)
	at org.h2.engine.Database.openDatabase(Database.java:270)
	at org.h2.engine.Database.<init>(Database.java:264)
	at org.h2.engine.Engine.openSession(Engine.java:64)
	at org.h2.engine.Engine.openSession(Engine.java:176)
	at org.h2.engine.Engine.createSessionAndValidate(Engine.java:154)
	at org.h2.engine.Engine.createSession(Engine.java:137)
	at org.h2.engine.Engine.createSession(Engine.java:27)
	at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:349)
	at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:118)
	at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:102)
	at org.h2.Driver.connect(Driver.java:72)
	at org.h2.jdbcx.JdbcDataSource.getJdbcConnection(JdbcDataSource.java:190)
	at org.h2.jdbcx.JdbcDataSource.getXAConnection(JdbcDataSource.java:353)
	at org.h2.jdbcx.JdbcDataSource.getPooledConnection(JdbcDataSource.java:385)
	at org.h2.jdbcx.JdbcConnectionPool.getConnectionNow(JdbcConnectionPool.java:227)
	at org.h2.jdbcx.JdbcConnectionPool.getConnection(JdbcConnectionPool.java:199)
	at org.eclipse.kura.core.db.H2DbServiceImpl.getConnection(H2DbServiceImpl.java:177)
	at org.eclipse.kura.core.db.H2DbServiceImpl.openDatabase(H2DbServiceImpl.java:321)
	at org.eclipse.kura.core.db.H2DbServiceImpl.openConnectionPool(H2DbServiceImpl.java:315)
	at org.eclipse.kura.core.db.H2DbServiceImpl.updated(H2DbServiceImpl.java:126)
	at org.eclipse.kura.core.db.H2DbServiceImpl.activate(H2DbServiceImpl.java:89)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.apache.felix.scr.impl.inject.BaseMethod.invokeMethod(BaseMethod.java:229)
	at org.apache.felix.scr.impl.inject.BaseMethod.access$500(BaseMethod.java:39)
	at org.apache.felix.scr.impl.inject.BaseMethod$Resolved.invoke(BaseMethod.java:650)
	at org.apache.felix.scr.impl.inject.BaseMethod.invoke(BaseMethod.java:506)
	at org.apache.felix.scr.impl.inject.ActivateMethod.invoke(ActivateMethod.java:307)
	at org.apache.felix.scr.impl.inject.ActivateMethod.invoke(ActivateMethod.java:299)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:298)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:109)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:906)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:879)
	at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:823)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse$1.run(ServiceFactoryUse.java:212)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.factoryGetService(ServiceFactoryUse.java:210)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.getService(ServiceFactoryUse.java:111)
	at org.eclipse.osgi.internal.serviceregistry.ServiceConsumer$2.getService(ServiceConsumer.java:45)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.getService(ServiceRegistrationImpl.java:508)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.getService(ServiceRegistry.java:461)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.getService(BundleContextImpl.java:624)
	at com.eclipsesource.jaxrs.publisher.internal.ResourceTracker.addingService(ResourceTracker.java:39)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:941)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:1)
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
	at org.osgi.util.tracker.AbstractTracked.trackInitial(AbstractTracked.java:183)
	at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:318)
	at org.osgi.util.tracker.ServiceTracker.open(ServiceTracker.java:261)
	at com.eclipsesource.jaxrs.publisher.internal.Activator.openAllServiceTracker(Activator.java:91)
	at com.eclipsesource.jaxrs.publisher.internal.Activator.start(Activator.java:55)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:779)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:1)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:772)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:729)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:933)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:309)
	at org.eclipse.osgi.container.Module.doStart(Module.java:581)
	at org.eclipse.osgi.container.Module.start(Module.java:449)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1634)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1614)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1585)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1528)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340)
Caused by: java.lang.IllegalStateException: File corrupted in chunk 15087, expected page length 4..768, got 200902975 [1.4.192/6]
	at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:773)
	at org.h2.mvstore.Page.read(Page.java:649)
	at org.h2.mvstore.Page.read(Page.java:195)
	at org.h2.mvstore.MVStore.readPage(MVStore.java:1944)
	at org.h2.mvstore.MVMap.readPage(MVMap.java:736)
	at org.h2.mvstore.Page.getChildPage(Page.java:217)
	at org.h2.mvstore.MVMap.binarySearch(MVMap.java:468)
	at org.h2.mvstore.MVMap.binarySearch(MVMap.java:469)
	at org.h2.mvstore.MVMap.get(MVMap.java:450)
	at org.h2.mvstore.MVStore.getChunkIfFound(MVStore.java:956)
	at org.h2.mvstore.MVStore.getChunk(MVStore.java:933)
	at org.h2.mvstore.MVStore.readPage(MVStore.java:1935)
	at org.h2.mvstore.MVMap.readPage(MVMap.java:736)
	at org.h2.mvstore.Page.getChildPage(Page.java:217)
	at org.h2.mvstore.MVMap.binarySearch(MVMap.java:468)
	at org.h2.mvstore.MVMap.binarySearch(MVMap.java:469)
	at org.h2.mvstore.MVMap.get(MVMap.java:450)
	at org.h2.mvstore.MVStore.getChunkIfFound(MVStore.java:956)
	at org.h2.mvstore.MVStore.getChunk(MVStore.java:933)
	at org.h2.mvstore.MVStore.readPage(MVStore.java:1935)
	at org.h2.mvstore.MVMap.readPage(MVMap.java:736)
	at org.h2.mvstore.Page.getChildPage(Page.java:217)
	at org.h2.mvstore.Cursor.min(Cursor.java:129)
	at org.h2.mvstore.Cursor.hasNext(Cursor.java:36)
	at org.h2.mvstore.MVStore.loadChunkMeta(MVStore.java:691)
	at org.h2.mvstore.MVStore.readStoreHeader(MVStore.java:672)
	at org.h2.mvstore.MVStore.<init>(MVStore.java:355)
	at org.h2.mvstore.MVStore$Builder.open(MVStore.java:2893)
	at org.h2.mvstore.db.MVTableEngine$Store.open(MVTableEngine.java:154)
	... 72 more
```

As of now, when this happens, the code in Kura simply drops the file and creates a new persistence file. Obviously, eventual data in the database will be lost.

```
2018-07-25 17:04:51,082 [Start Level: Equinox Container: e3729c13-30bd-4b97-9aa6-783d41607f14] WARN  o.e.k.c.d.H2DbServiceImpl - Deleting database files...
2018-07-25 17:04:51,104 [Start Level: Equinox Container: e3729c13-30bd-4b97-9aa6-783d41607f14] WARN  o.e.k.c.d.H2DbServiceImpl - Deleting database files...done
2018-07-25 17:04:52,063 [Start Level: Equinox Container: e3729c13-30bd-4b97-9aa6-783d41607f14] INFO  o.e.k.c.d.H2DbServiceImpl - updating...done
2018-07-25 17:04:52,064 [Start Level: Equinox Container: e3729c13-30bd-4b97-9aa6-783d41607f14] INFO  o.e.k.c.d.H2DbServiceImpl - activating...done
```

We should change the activation code to try to recover the database file and save as many entrie in the database as possible.

I've also discovered another problem with the database: it starts well, stores messages without error, but when a query is executed an exception like this is thrown:

```
2018-07-25 17:01:55,032 [HouseKeeperTask] WARN  o.e.k.c.d.s.HouseKeeperTask - HouseCleaningTask exception
org.eclipse.kura.KuraStoreException: "Error performing operation on store. Cannot execute query"
	at org.eclipse.kura.core.data.store.DbDataStore.executeDeleteMessagesQuery(DbDataStore.java:597)
	at org.eclipse.kura.core.data.store.DbDataStore.deleteStaleMessages(DbDataStore.java:446)
	at org.eclipse.kura.core.data.store.HouseKeeperTask.run(HouseKeeperTask.java:53)
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.util.concurrent.FutureTask.runAndReset(Unknown Source)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(Unknown Source)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
Caused by: org.h2.jdbc.JdbcSQLException: General error: "java.lang.IllegalStateException: File corrupted in chunk 14820, expected page length 4..512, got -397012734 [1.4.192/6]"; SQL statement:
DELETE FROM ds_messages WHERE qos = 0 AND publishedOn <= DATEADD('ss', -?, ?) AND publishedOn IS NOT NULL; [50000-192]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:168)
	at org.h2.message.DbException.convert(DbException.java:295)
	at org.h2.command.Command.executeUpdate(Command.java:266)
	at org.h2.jdbc.JdbcPreparedStatement.execute(JdbcPreparedStatement.java:201)
	at org.eclipse.kura.core.data.store.DbDataStore.executeDeleteMessagesQuery(DbDataStore.java:593)
	... 9 more
Caused by: java.lang.IllegalStateException: File corrupted in chunk 14820, expected page length 4..512, got -397012734 [1.4.192/6]
	at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:773)
	at org.h2.mvstore.Page.read(Page.java:649)
	at org.h2.mvstore.Page.read(Page.java:195)
	at org.h2.mvstore.MVStore.readPage(MVStore.java:1944)
	at org.h2.mvstore.MVMap.readPage(MVMap.java:736)
	at org.h2.mvstore.Page.getChildPage(Page.java:217)
	at org.h2.mvstore.MVMap.binarySearch(MVMap.java:468)
	at org.h2.mvstore.MVMap.binarySearch(MVMap.java:469)
	at org.h2.mvstore.MVMap.get(MVMap.java:450)
	at org.h2.mvstore.db.TransactionStore$TransactionMap.getValue(TransactionStore.java:1203)
	at org.h2.mvstore.db.TransactionStore$TransactionMap.get(TransactionStore.java:1180)
	at org.h2.mvstore.db.TransactionStore$TransactionMap.get(TransactionStore.java:1148)
	at org.h2.mvstore.db.MVPrimaryIndex.getRow(MVPrimaryIndex.java:211)
	at org.h2.mvstore.db.MVTable.getRow(MVTable.java:463)
	at org.h2.mvstore.db.MVSecondaryIndex$MVStoreCursor.get(MVSecondaryIndex.java:489)
	at org.h2.index.IndexCursor.get(IndexCursor.java:288)
	at org.h2.table.TableFilter.get(TableFilter.java:593)
	at org.h2.command.dml.Delete.update(Delete.java:75)
	at org.h2.command.CommandContainer.update(CommandContainer.java:98)
	at org.h2.command.Command.executeUpdate(Command.java:258)
	... 11 more
```

The file seems to be corrupted, but the activation didn't discover it and all the query operations cannot be performed.

I believe we should be able to address this case too.

--- a/_posts/builtin/2017-10-18-h2-database.md
+++ b/_posts/builtin/2017-10-18-h2-database.md
@@ -21,7 +21,7 @@ Starting from version 3.1.0, Eclipse Kura includes the [H2 database engine](http
 
 Eclipse Kura supports the following H2 database features:
 
-* **Persistence modes**: The H2 implementation currently supports in-memory and file-based database instances. See 
+* **Persistence modes**: The H2 implementation currently supports in-memory and file-based database instances. See
 [Persistence Modes](#persistence-modes) for more details.
 
 * **Multiple database instances**: It is possible to create and configure multiple database instances from the Kura Web UI, these instances can be selectively consumed by applications. A default database instance is created automatically.
@@ -142,7 +142,7 @@ In order to enable the H2 Web console, proceed as follows:
 
 #### Changing the password
 
-The prerequisite for being able to change the password is that 
+The prerequisite for being able to change the password is that
 
 1. Open the configuration of the desired database instance in the Kura Web UI.
 
@@ -174,11 +174,11 @@ A persistent database instance can be created using the `jdbc:h2:file:<dbpath>`,
 If no URL parameters are supplied the database will enable the transaction log by default. The transaction log is used to restore the database to a consistent state after a crash or power failure. This provides good protection against data losses but causes a lot of writes to the storage device, reducing both performance and the lifetime of flash-based storage devices.
 
 This configuration is suggested for database instances that are rarely updated.
-Examples: 
+Examples:
 
 * `jdbc:h2:file:/opt/db/mydb`
 
-Make sure to use absolute paths in the db URL since H2 does not support db paths relative to the working directory
+Make sure to use absolute paths in the db URL since H2 does not support db paths relative to the working directory.
 
 #### Mostly persistent
 
@@ -187,6 +187,47 @@ The transaction log can be disabled by appending the `LOG=0` parameter to the DB
 In order to reduce the probability of data losses the H2DbService performs periodic checkpoints on the database. A checkpoint forces all pending modifications to be committed to the storage device. The interval in seconds between two consecutive checkpoints can be configured using the **db.checkpoint.interval.seconds** property.
 
 This configuration is suggested for database instances with intermediate update rates.
-Examples: 
+Examples:
 
 * `jdbc:h2:file:/opt/db/mydb;LOG=0`
+
+#### Filesystem tuning
+
+The H2 database might become corrupted [1] in case of power failure.
+In case of file-backed database (e.g. mostly or most persistent), the following database mount options can improve database reliability, reducing the database corruption probability at the expense of reduced performance and lifetime of the flash storage devices:
+
+* `data=journal`: From EXT4(5) Linux MAN Page:
+
+  > All data is committed into the journal prior to being written into the main filesystem.
+
+  This setting only applies to ext3/ext4 filesystems.
+
+* `sync`: From mount(8) Linux MAN Page:
+
+  > All I/O to the filesystem should be done synchronously. In case of media with limited number of write cycles (e.g. some flash drives) "sync" may cause life-cycle shortening.
+
+After some long running tests [1], it appears that the `data=journal` option alone is enough to reduce corruption probability (no corruption observed during the test).
+
+[1] https://github.com/eclipse/kura/issues/2169
+
+In order to apply the options above, perform the following steps:
+
+1 - Edit the `/etc/fstab` row for the filesystem containing the database file, appending the desired options to the fourth column, using comma as separator.
+
+For example on Raspbian an `/etc/fstab` file with `data=journal` enabled will look like the following:
+
+```
+proc            /proc           proc    defaults          0       0
+PARTUUID=3920f25c-01  /boot           vfat    defaults          0       2
+PARTUUID=3920f25c-02  /               ext4    defaults,noatime,data=journal  0       1
+# a swapfile is not a swap partition, no line here
+#   use  dphys-swapfile swap[on|off]  for that
+```
+
+2 - If the database resides on the root filesystem, the `data=journal` option must be specified in the kernel command line as well, by appending `rootflags=data=journal`
+
+On Raspbian this can be done by editing the `/boot/cmdline.txt` file:
+
+```
+dwc_otg.lpm_enable=0 console=serial0,115200 console=tty1 root=PARTUUID=3920f25c-02 rootfstype=ext4 elevator=deadline fsck.repair=yes rootwait rootflags=data=journal
+```

